### PR TITLE
feat: Add anonymized signature metrics

### PR DESCRIPTION
**Pull request by pnarayanaswamy**
## **Description**

Adds logic to send Signature Requested Anon, Signature Approved Anon and Signature Rejected Anon events when the signature is of type signTypedData_v4 with additional anonymous properties.
Updates the e2e to check the correct Anon events are sent with the additional properties
Added additional unit tests to increase sonar coverage
Exclude the test folder from sonar checks

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27162?quickstart=1)

## **Related issues**

Fixes: [3009](https://github.com/MetaMask/MetaMask-planning/issues/3009)

## **Manual testing steps**

1. Go to test dapp
2. Under the Permit section, click on Sign
3. This should trigger the Signature Anon events with:
      ```
      eip712_verifyingContract == 0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC
      eip712_domain_version == 1
      eip712_domain_name == MyToken
      ```

1. Go to test dapp
2. Under the SignTypedData_v4 section, click on Sign
3. This should trigger the Signature Anon events with:
      ```
      eip712_verifyingContract == 0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC
      eip712_domain_version == 1
      eip712_domain_name == Ether Mail
      ```

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/controllers/metametrics.test.js [hunk 1]
```diff
@@ -580,6 +580,77 @@ describe('MetaMetricsController', function () {
...
+        event: `Signature Approved Anon`,
+        properties: { foo: 'bar', ...DEFAULT_EVENT_PROPERTIES },
+      });
+
+      expect(spy.mock.calls[1][0]).toMatchObject({
+        event: `Signature Approved`,
+        properties: { ...DEFAULT_EVENT_PROPERTIES },
+      });
+    });
+  });
```
Comment (digiwand) [inline app/scripts/controllers/metametrics.test.js:652]: [non-blocking] we could DRY these up by looping the tests together

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js [hunk 2]
```diff
@@ -222,8 +224,7 @@ export default function createRPCMethodTrackingMiddleware({
   ) {
     const { origin, method, params } = req;
 
-    const rateLimitType =
-      RATE_LIMIT_MAP[method] ?? RATE_LIMIT_TYPES.RANDOM_SAMPLE;
```
Comment (digiwand) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:226]: why did we remove the `RATE_LIMIT_TYPES.RANDOM_SAMPLE` fallback?

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js [hunk 3]
```diff
@@ -346,14 +346,25 @@ export default function createRPCMethodTrackingMiddleware({
...
+            const parsedMessageData = parseTypedDataMessage(data);
+            sensitiveEventProperties = {};
+
+            eventProperties.eip712_primary_type = parsedMessageData.primaryType;
+            sensitiveEventProperties.eip712_verifyingContract =
+              parsedMessageData.domain.verifyingContract;
+            sensitiveEventProperties.eip712_domain_version =
+              parsedMessageData.domain.version;
+            sensitiveEventProperties.eip712_domain_name =
+              parsedMessageData.domain.name;
```
Comment (digiwand) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:353]: could do something like this. Would need to update primaryType again if this suggestion is accepted 

```suggestion
            const { domain, primaryType } = parseTypedDataMessage(data);

            sensitiveEventProperties = {
              eip712_verifyingContract: domain.verifyingContract,
              eip712_domain_version: domain.version,
              eip712_domain_name: domain.name,
            };

            eventProperties.eip712_primary_type = primaryType;
```

File: app/scripts/controllers/metametrics.test.js (hunk 1 above)
Comment (pnarayanaswamy) [reply app/scripts/controllers/metametrics.test.js:613]: done

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js (hunk 2 above)
Comment (pnarayanaswamy) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js:226]: the switch has a default condition that falls back to the RANDOM_SAMPLE anyhow, so this is redundant

**Review by digiwand — APPROVED** [review]

**Review by jpuri — APPROVED** [review]

**Pull request merged**